File: packages/edge/src/middleware/ip-filter.ts
import { Request, Response, NextFunction } from 'express';
import { createLogger } from '../config/logger';
import net from 'net';

const logger = createLogger('IpFilter');

interface CidrRange {
  addr: number[];
  prefixLen: number;
  isV6: boolean;
}

/**
 * Parse a CIDR string (e.g., "10.0.0.0/8" or "192.168.1.1")
 */
function parseCidr(s: string): CidrRange | null {
  const trimmed = s.trim();
  if (!trimmed) return null;

  const [addrStr, prefixStr] = trimmed.split('/');

  if (net.isIPv4(addrStr)) {
    const parts = addrStr.split('.').map(Number);
    if (parts.some((p) => isNaN(p) || p < 0 || p > 255)) return null;
    const prefixLen = prefixStr ? parseInt(prefixStr, 10) : 32;
    if (isNaN(prefixLen) || prefixLen < 0 || prefixLen > 32) return null;
    return { addr: parts, prefixLen, isV6: false };
  }

  // IPv6 is less common for Edge, skip for now
  return null;
}

/**
 * Check if an IPv4 address matches a CIDR range
 */
function cidrContains(cidr: CidrRange, ip: string): boolean {
  if (cidr.isV6) return false;

  const ipParts = ip.split('.').map(Number);
  if (ipParts.length !== 4 || ipParts.some((p) => isNaN(p))) return false;

  // Convert to 32-bit integers
  const ipNum =
    (ipParts[0] << 24) | (ipParts[1] << 16) | (ipParts[2] << 8) | ipParts[3];
  const cidrNum =
    (cidr.addr[0] << 24) |
    (cidr.addr[1] << 16) |
    (cidr.addr[2] << 8) |
    cidr.addr[3];

  if (cidr.prefixLen === 0) return true;
  const mask = ~0 << (32 - cidr.prefixLen);
  return (ipNum & mask) === (cidrNum & mask);
}

/**
 * Parse a comma-separated list of CIDRs
 */
function parseCidrList(s: string): CidrRange[] {
  if (!s) return [];
  return s
    .split(',')
    .map((part) => parseCidr(part))
    .filter((c): c is CidrRange => c !== null);
}

/**
 * Normalize IP address (strip IPv6 prefix like ::ffff:)
 */
function normalizeIp(ip: string): string {
  if (ip.startsWith('::ffff:')) {
    return ip.substring(7);
  }
  return ip;
}

/**
 * Create an IP filter middleware.
 * When both lists are empty, the middleware is a pass-through.
 */
export function createIpFilter(allowIps: string, denyIps: string) {
  const allowList = parseCidrList(allowIps);
  const denyList = parseCidrList(denyIps);

  if (allowList.length > 0) {
    logger.info(`IP allow list configured with ${allowList.length} entries`);
  }
  if (denyList.length > 0) {
    logger.info(`IP deny list configured with ${denyList.length} entries`);
  }

  return (req: Request, res: Response, next: NextFunction): void => {
    if (allowList.length === 0 && denyList.length === 0) {
      next();
      return;
    }

    const rawIp = req.ip || req.socket.remoteAddress || '127.0.0.1';
    const clientIp = normalizeIp(rawIp);

    // Check deny list first
    if (
      denyList.length > 0 &&
      denyList.some((cidr) => cidrContains(cidr, clientIp))
    ) {
      logger.warn(`IP denied by deny list: ${clientIp}`);
      res.status(403).json({
        success: false,
        error: {
          code: 'FORBIDDEN',
          message: 'Access denied',
        },
      });
      return;
    }

    // Check allow list
    if (
      allowList.length > 0 &&
      !allowList.some((cidr) => cidrContains(cidr, clientIp))
    ) {
      logger.warn(`IP not in allow list: ${clientIp}`);
      res.status(403).json({
        success: false,
        error: {
          code: 'FORBIDDEN',
          message: 'Access denied',
        },
      });
      return;
    }

    next();
  };
}

File: packages/edge-rust/src/middleware/ip_filter.rs
// Gatrix Edge Rust - IP Filter Middleware
// IP allow/deny list with CIDR support

use std::net::IpAddr;
use actix_web::dev::{Service, ServiceRequest, ServiceResponse, Transform};
use actix_web::HttpResponse;
use futures_util::future::{ok, Either, Ready};
use std::future::Future;
use std::pin::Pin;
use std::task::{Context, Poll};

/// CIDR network range
#[derive(Clone, Debug)]
pub struct CidrRange {
  addr: IpAddr,
  prefix_len: u8,
}

impl CidrRange {
  /// Parse a CIDR string (e.g., "10.0.0.0/8" or "192.168.1.1")
  pub fn parse(s: &str) -> Option<Self> {
    let s = s.trim();
    if let Some((addr_str, prefix_str)) = s.split_once('/') {
      let addr: IpAddr = addr_str.parse().ok()?;
      let prefix_len: u8 = prefix_str.parse().ok()?;
      Some(Self { addr, prefix_len })
    } else {
      // Single IP, treat as /32 or /128
      let addr: IpAddr = s.parse().ok()?;
      let prefix_len = match addr {
        IpAddr::V4(_) => 32,
        IpAddr::V6(_) => 128,
      };
      Some(Self { addr, prefix_len })
    }
  }

  /// Check if an IP address is within this CIDR range
  pub fn contains(&self, ip: &IpAddr) -> bool {
    match (&self.addr, ip) {
      (IpAddr::V4(net), IpAddr::V4(target)) => {
        let net_bits = u32::from(*net);
        let target_bits = u32::from(*target);
        if self.prefix_len == 0 {
          return true;
        }
        let mask = !0u32 << (32 - self.prefix_len);
        (net_bits & mask) == (target_bits & mask)
      }
      (IpAddr::V6(net), IpAddr::V6(target)) => {
        let net_bits = u128::from(*net);
        let target_bits = u128::from(*target);
        if self.prefix_len == 0 {
          return true;
        }
        let mask = !0u128 << (128 - self.prefix_len);
        (net_bits & mask) == (target_bits & mask)
      }
      _ => false, // Mixed v4/v6, no match
    }
  }
}

/// Parse comma-separated CIDR list
pub fn parse_cidr_list(s: &str) -> Vec<CidrRange> {
  s.split(',')
    .filter_map(|part| CidrRange::parse(part.trim()))
    .collect()
}

/// IP filter middleware
#[derive(Clone)]
pub struct IpFilter {
  allow_list: Vec<CidrRange>,
  deny_list: Vec<CidrRange>,
}

impl IpFilter {
  pub fn new(allow_list: Vec<CidrRange>, deny_list: Vec<CidrRange>) -> Self {
    Self { allow_list, deny_list }
  }
}

impl<S, B> Transform<S, ServiceRequest> for IpFilter
where
  S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error> + 'static,
  B: 'static,
{
  type Response = ServiceResponse<actix_web::body::EitherBody<B>>;
  type Error = actix_web::Error;
  type Transform = IpFilterMiddleware<S>;
  type InitError = ();
  type Future = Ready<Result<Self::Transform, Self::InitError>>;

  fn new_transform(&self, service: S) -> Self::Future {
    ok(IpFilterMiddleware {
      service,
      allow_list: self.allow_list.clone(),
      deny_list: self.deny_list.clone(),
    })
  }
}

pub struct IpFilterMiddleware<S> {
  service: S,
  allow_list: Vec<CidrRange>,
  deny_list: Vec<CidrRange>,
}

impl<S, B> Service<ServiceRequest> for IpFilterMiddleware<S>
where
  S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error> + 'static,
  B: 'static,
{
  type Response = ServiceResponse<actix_web::body::EitherBody<B>>;
  type Error = actix_web::Error;
  type Future = Either<
    Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>,
    Ready<Result<Self::Response, Self::Error>>,
  >;

  fn poll_ready(&self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
    self.service.poll_ready(cx)
  }

  fn call(&self, req: ServiceRequest) -> Self::Future {
    let client_ip_str = req
      .connection_info()
      .realip_remote_addr()
      .unwrap_or("127.0.0.1")
      .to_string();

    // Strip port from IP if present (e.g., "127.0.0.1:12345")
    let ip_only = client_ip_str
      .rsplit_once(':')
      .map(|(ip, _port)| ip)
      .unwrap_or(&client_ip_str);

    let client_ip: Option<IpAddr> = ip_only.parse().ok();

    if let Some(ip) = client_ip {
      // Check deny list first
      if !self.deny_list.is_empty() && self.deny_list.iter().any(|cidr| cidr.contains(&ip)) {
        log::warn!("IP denied by deny list: {}", ip);
        let response = HttpResponse::Forbidden().json(serde_json::json!({
          "success": false,
          "error": {
            "code": "FORBIDDEN",
            "message": "Access denied"
          }
        }));
        return Either::Right(ok(req.into_response(response).map_into_right_body()));
      }

      // Check allow list
      if !self.allow_list.is_empty() && !self.allow_list.iter().any(|cidr| cidr.contains(&ip)) {
        log::warn!("IP not in allow list: {}", ip);
        let response = HttpResponse::Forbidden().json(serde_json::json!({
          "success": false,
          "error": {
            "code": "FORBIDDEN",
            "message": "Access denied"
          }
        }));
        return Either::Right(ok(req.into_response(response).map_into_right_body()));
      }
    }

    let fut = self.service.call(req);
    Either::Left(Box::pin(async move {
      let res = fut.await?;
      Ok(res.map_into_left_body())
    }))
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_cidr_parse() {
    let cidr = CidrRange::parse("10.0.0.0/8").unwrap();
    assert!(cidr.contains(&"10.1.2.3".parse().unwrap()));
    assert!(cidr.contains(&"10.255.255.255".parse().unwrap()));
    assert!(!cidr.contains(&"11.0.0.1".parse().unwrap()));
  }

  #[test]
  fn test_single_ip() {
    let cidr = CidrRange::parse("192.168.1.1").unwrap();
    assert!(cidr.contains(&"192.168.1.1".parse().unwrap()));
    assert!(!cidr.contains(&"192.168.1.2".parse().unwrap()));
  }

  #[test]
  fn test_cidr_list_parse() {
    let list = parse_cidr_list("10.0.0.0/8, 192.168.0.0/16, 172.16.0.1");
    assert_eq!(list.len(), 3);
  }
}
